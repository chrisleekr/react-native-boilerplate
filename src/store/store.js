import {configureStore} from '@reduxjs/toolkit';
import {combineReducers} from 'redux';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER
} from 'redux-persist';
import AsyncStorage from '@react-native-async-storage/async-storage';

// Axios
import setupInterceptors from '../service/setupInterceptors';

// Slices
import authReducer from './slice/authSlice';
import todoReducer from './slice/todoSlice';
import userReducer from './slice/userSlice';

const persistConfig = {
  key: 'root',
  version: 1,
  storage: AsyncStorage
};

const rootReducer = combineReducers({
  auth: authReducer,
  todo: todoReducer,
  user: userReducer
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

const store = configureStore({
  reducer: persistedReducer,
  // middleware option needs to be provided for avoiding the error. ref: https://redux-toolkit.js.org/usage/usage-guide#use-with-redux-persist
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware({
      // Since axios is throwing serializecheck error, so had to turn off.
      // serializableCheck: false
      // But let see how it goes with following list instead of turnning it off.
      serializableCheck: {
        ignoredActions: [
          FLUSH,
          REHYDRATE,
          PAUSE,
          PERSIST,
          PURGE,
          REGISTER,
          'auth/login/rejected',
          'auth/register/rejected',
          'auth/passwordResetRequest/rejected',
          'todo/list/rejected',
          'todo/addOne/rejected',
          'todo/updateOne/rejected',
          'todo/deleteOne/rejected',
          'user/me/rejected',
          'user/updateMe/rejected'
        ]
      }
    })
});

setupInterceptors(store);

export const persistor = persistStore(store);
export default store;
