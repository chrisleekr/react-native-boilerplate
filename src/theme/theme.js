import {Dimensions} from 'react-native';
import {colors} from './colors';
const {width, height} = Dimensions.get('window');

// Themes:- Can alter values here. Can only be consumed through Context (see useTheme.js file)
const themes = {
  light: {
    name: 'light',
    barStyle: 'light-content',
    color: '#000000',
    labelColor: '#525E75',
    invertColor: '#ffffff',
    layoutBg: '#f0f0f0',
    headerBg: '#f6f7f8',
    cardBg: '#e4dfa7',
    accent: '#0071ff',
    error: '#B00020',
    errorText: '#ed4337',
    lightBg: '#F5EE9E',
    borderColor: '#B00020',
    buttonActiveColor: '#6200EE',
    buttonDisabledColor: 'rgba(0, 0, 0, 0.26)',
    card: {
      shadowColor: '#000000',
      borderColor: '#B00020'
    },
    ...colors
  },
  dark: {
    name: 'dark',
    barStyle: 'dark-content',
    color: '#ffffff',
    labelColor: '#F1DDBF',
    invertColor: '#000000',
    layoutBg: '#121212',
    headerBg: '#1e1e1e',
    cardBg: '#3f4045',
    accent: '#0071ff',
    error: '#B00020',
    errorText: '#ed4337',
    lightBg: '#F5EE9E',
    borderColor: '#ffffff',
    buttonActiveColor: '#BB86FC',
    buttonDisabledColor: 'rgba(255,255,255, 0.26)',
    card: {
      shadowColor: '#121212',
      borderColor: '#ffffff'
    },
    ...colors
  }
};

const window = {
  // app dimensions
  width,
  height
};

export {themes, window};
