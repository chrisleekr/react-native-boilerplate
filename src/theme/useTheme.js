/**
 * For Dark/Light Mode
 * @exports ThemeContext, ThemeProvider, useTheme
 */

import React, {useState, createContext, useContext} from 'react';
import {Appearance} from 'react-native';

// Import preconfigured themes from theme file
import {themes} from './theme';

// Context
export const ThemeContext = createContext(null);

// Provider to be used in index/App/or top of any parent
export const ThemeProvider = ({children}) => {
  const colorScheme = Appearance.getColorScheme();

  const [theme, setTheme] = useState(themes[colorScheme]);
  return (
    <ThemeContext.Provider value={{theme, setTheme}}>
      {children}
    </ThemeContext.Provider>
  );
};

// useTheme hook for easy access to theme and setTheme
export const useTheme = () => {
  const {theme, setTheme} = useContext(ThemeContext);

  const toggleTheme = v => {
    setTheme(v ? themes.dark : themes.light);
  };

  return {theme, toggleTheme};
};
