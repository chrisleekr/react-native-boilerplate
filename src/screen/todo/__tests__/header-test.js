import React from 'react';
import * as redux from 'react-redux';
import {NavigationContainer} from '@react-navigation/native';
import {
  cleanup,
  fireEvent,
  render,
  waitFor,
  act,
  within
} from '@testing-library/react-native';

import {ThemeProvider} from '../../../theme/useTheme';

import store from '../../../store/store';

import TodoHeader from '../header';

let rendered;

let mockList = Promise.resolve({
  success: true,
  status: 200,
  message: '',
  data: {
    rows: [
      {
        id: 1,
        name: '123456',
        state: 'ongoing'
      }
    ],
    pagination: {page: 1, page_size: 999999, total_rows: 3, first_row_no: 3}
  }
});

let mockPostOne;

jest.mock('../../../service/todoService', () => ({
  list: () => mockList,
  postOne: () => mockPostOne
}));

const component = (
  <ThemeProvider>
    <redux.Provider store={store}>
      <NavigationContainer>
        <TodoHeader />
      </NavigationContainer>
    </redux.Provider>
  </ThemeProvider>
);

describe('TodoHeader', () => {
  afterEach(cleanup);

  beforeEach(async () => {
    rendered = render(component);
    await waitFor(() => expect(rendered.toJSON()).not.toBeNull());
  });

  it('matches snapshot', () => {
    expect(rendered.toJSON()).toMatchSnapshot();
  });

  describe('clicks register button without entering fields', () => {
    beforeEach(async () => {
      fireEvent.press(
        within(rendered.getByTestId('name-input')).getByTestId('icon')
      );
      await waitFor(() => expect(rendered.toJSON()).not.toBeNull());
    });

    it('matches snapshot', () => {
      expect(rendered.toJSON()).toMatchSnapshot();
    });
  });

  describe('success', () => {
    it('enters todo', async () => {
      act(() => {
        fireEvent.changeText(
          within(rendered.getByTestId('name-input')).getByTestId('input'),
          'Lee'
        );
      });

      fireEvent.press(
        within(rendered.getByTestId('name-input')).getByTestId('icon')
      );

      await waitFor(() => expect(rendered.toJSON()).not.toBeNull());
    });
  });

  describe('fail', () => {
    it('enters todo', async () => {
      mockPostOne = Promise.reject({
        response: {
          status: 500,
          data: {
            data: [{msg: 'Error'}]
          }
        }
      });

      act(() => {
        fireEvent.changeText(
          within(rendered.getByTestId('name-input')).getByTestId('input'),
          'Lee'
        );
      });

      fireEvent.press(
        within(rendered.getByTestId('name-input')).getByTestId('icon')
      );

      await waitFor(() => expect(rendered.toJSON()).not.toBeNull());
    });
  });
});
