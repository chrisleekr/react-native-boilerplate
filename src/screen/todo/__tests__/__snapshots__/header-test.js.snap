// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoHeader clicks register button without entering fields matches snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "borderBottomLeftRadius": 30,
        "borderBottomRightRadius": 30,
        "height": 85,
        "paddingHorizontal": 10,
        "paddingTop": 10,
        "zIndex": 1,
      },
      Object {
        "backgroundColor": "#f6f7f8",
      },
    ]
  }
  testID="todo-header"
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 13,
        },
        Object {
          "color": "#000000",
        },
      ]
    }
  >
    What do you need to do?
  </Text>
  <View
    style={
      Array [
        Object {
          "flex": 1,
          "flexDirection": "row",
        },
        Object {
          "marginHorizontal": 0,
          "marginTop": 5,
        },
      ]
    }
    testID="name-input"
  >
    <TextInput
      onBlur={[Function]}
      onChangeText={[Function]}
      style={
        Array [
          Object {
            "borderWidth": 1,
            "height": 35,
            "paddingBottom": 0,
            "paddingTop": 0,
            "textAlignVertical": "center",
            "width": "100%",
          },
          Array [
            Object {
              "borderRadius": 5,
              "borderWidth": 0,
            },
            Object {
              "backgroundColor": "#000000",
            },
          ],
          Object {
            "borderColor": "red",
            "borderWidth": 1,
          },
          Object {
            "backgroundColor": "#e4dfa7",
          },
        ]
      }
      testID="input"
      value=""
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
          "position": "absolute",
          "right": 0,
          "textAlignVertical": "center",
          "top": 2,
          "zIndex": 1,
        }
      }
      testID="icon"
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#000000",
              "fontSize": 30,
            },
            undefined,
            Object {
              "fontFamily": "Material Design Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        󰐕
      </Text>
    </View>
  </View>
</View>
`;

exports[`TodoHeader matches snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "borderBottomLeftRadius": 30,
        "borderBottomRightRadius": 30,
        "height": 85,
        "paddingHorizontal": 10,
        "paddingTop": 10,
        "zIndex": 1,
      },
      Object {
        "backgroundColor": "#f6f7f8",
      },
    ]
  }
  testID="todo-header"
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 13,
        },
        Object {
          "color": "#000000",
        },
      ]
    }
  >
    What do you need to do?
  </Text>
  <View
    style={
      Array [
        Object {
          "flex": 1,
          "flexDirection": "row",
        },
        Object {
          "marginHorizontal": 0,
          "marginTop": 5,
        },
      ]
    }
    testID="name-input"
  >
    <TextInput
      onBlur={[Function]}
      onChangeText={[Function]}
      style={
        Array [
          Object {
            "borderWidth": 1,
            "height": 35,
            "paddingBottom": 0,
            "paddingTop": 0,
            "textAlignVertical": "center",
            "width": "100%",
          },
          Array [
            Object {
              "borderRadius": 5,
              "borderWidth": 0,
            },
            Object {
              "backgroundColor": "#000000",
            },
          ],
          undefined,
          Object {
            "backgroundColor": "#e4dfa7",
          },
        ]
      }
      testID="input"
      value=""
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
          "position": "absolute",
          "right": 0,
          "textAlignVertical": "center",
          "top": 2,
          "zIndex": 1,
        }
      }
      testID="icon"
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#000000",
              "fontSize": 30,
            },
            undefined,
            Object {
              "fontFamily": "Material Design Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        󰐕
      </Text>
    </View>
  </View>
</View>
`;
